pkg/ecdsa/tss/protocol_recovery.go

```diff
@@ -0,0 +1,147 @@
+import (
+	"context"
+	cecdsa "crypto/ecdsa"
+	"fmt"
+	"time"
+
+	"github.com/keep-network/keep-core/pkg/net"
+)
+
+// BroadcastRecoveryAddress broadcasts and receives the BTC recovery addresses
+// of each client so that each client can retrieve the underlying bitcoin in
+// the case that a keep is terminated.
+func broadcastRecoveryAddress(
+	parentCtx context.Context,
+	groupID string,
+	memberID MemberID,
+	groupMemberIDs []MemberID,
+	dishonestThreshold uint,
+	networkProvider net.Provider,
+	pubKeyToAddressFn func(cecdsa.PublicKey) []byte,
+) error {
+	const protocolReadyTimeout = 2 * time.Minute
+
+	group := &groupInfo{
+		groupID:            groupID,
+		memberID:           memberID,
+		groupMemberIDs:     groupMemberIDs,
+		dishonestThreshold: int(dishonestThreshold),
+	}
+
+	netBridge, _ := newNetworkBridge(group, networkProvider)
+	broadcastChannel, _ := netBridge.getBroadcastChannel()
+	ctx, cancel := context.WithTimeout(parentCtx, protocolReadyTimeout)
+	defer cancel()
+
+	msgInChan := make(chan *LiquidationRecoveryMessage, len(group.groupMemberIDs))
+	handleLiquidationRecoveryMessage := func(netMsg net.Message) {
+		switch msg := netMsg.Payload().(type) {
+		case *LiquidationRecoveryMessage:
+			msgInChan <- msg
+		}
+	}
+	broadcastChannel.Recv(ctx, handleLiquidationRecoveryMessage)
+
+	memberBTCAddresses := make(map[string]string)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case msg := <-msgInChan:
+				for _, memberID := range group.groupMemberIDs {
+					if msg.SenderID.Equal(memberID) {
+						memberAddress, err := memberIDToAddress(
+							memberID,
+							pubKeyToAddressFn,
+						)
+						if err != nil {
+							logger.Errorf(
+								"could not convert member ID to address for "+
+									"a member of keep [%s]: [%v]",
+								group.groupID,
+								err,
+							)
+							break
+						}
+						memberBTCAddresses[memberAddress] = msg.BtcRecoveryAddress
+
+						logger.Infof(
+							"member [%s] from keep [%s] announced supplied btc address [%s] for liquidation recovery",
+							memberAddress,
+							group.groupID,
+							msg.BtcRecoveryAddress,
+						)
+
+						break
+					}
+				}
+
+				if len(memberBTCAddresses) == len(group.groupMemberIDs) {
+					cancel()
+				}
+			}
+		}
+	}()
+
+	go func() {
+		sendMessage := func() {
+			if err := broadcastChannel.Send(ctx,
+				&LiquidationRecoveryMessage{SenderID: group.memberID, BtcRecoveryAddress: "abc123"},
+			); err != nil {
+				logger.Errorf("failed to send btc recovery address: [%v]", err)
+			}
+		}
+
+		// Send the message first time. It will be periodically retransmitted
+		// by the broadcast channel for the entire lifetime of the context.
+		sendMessage()
+
+		<-ctx.Done()
+		// Send the message once again as the member received messages
+		// from all peer members but not all peer members could receive
+		// the message from the member as some peer member could join
+		// the protocol after the member sent the last message.
```

Comment:
Hmmm… Not sure this is entirely clear to me, but more to the point---this seems like something we should really wrap up in a helper if we're doing it in every protocol. Not necessarily something we want to tackle in liquidation recovery though.